File: Heat/Extensions/EnumerableExtensions.vb
Imports System.Runtime.CompilerServices

Public Module EnumerableExtensions

    <Extension> _
    Public Function ToSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValue As String) As IEnumerable(Of SelectListItem)
        Return items.ToSelectListItems(name, value, selectedValue, False)
    End Function

    <Extension> _
    Public Function ToSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValue As String, includeNotApplicable As Boolean, Optional notApplicableValue As String = "", _
        Optional notApplicableText As String = "(Not Applicable)") As IEnumerable(Of SelectListItem)
        Return items.ToSelectListItems(name, value, Function(x) value(x) = selectedValue, includeNotApplicable, notApplicableValue, notApplicableText)
    End Function

    <Extension> _
    Public Function ToSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), isSelected As Func(Of T, Boolean)) As IEnumerable(Of SelectListItem)
        Return items.ToSelectListItems(name, value, isSelected, False)
    End Function

    <Extension> _
    Public Function ToSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), isSelected As Func(Of T, Boolean), includeNotApplicable As Boolean, Optional notApplicableValue As String = "", _
        Optional notApplicableText As String = "(Not Applicable)") As IEnumerable(Of SelectListItem)
        If includeNotApplicable Then
            Dim returnItems = New List(Of SelectListItem)() From { _
                New SelectListItem() With { _
                     .Text = notApplicableText, _
                     .Value = notApplicableValue, _
                     .Selected = False _
                } _
            }

            If items IsNot Nothing Then
                returnItems.AddRange(items.[Select](Function(item) New SelectListItem() With { _
                    .Text = name(item), _
                     .Value = value(item), _
                     .Selected = isSelected(item) _
                }))
            End If
            Return returnItems
        End If

        If items Is Nothing Then
            Return New List(Of SelectListItem)()
        End If

        Return items.[Select](Function(item) New SelectListItem() With { _
             .Text = name(item), _
             .Value = value(item), _
             .Selected = isSelected(item) _
        })
    End Function

    <Extension> _
    Public Function ToMultiSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValues As IEnumerable(Of Integer)) As IEnumerable(Of SelectListItem)
        If selectedValues Is Nothing Then
            selectedValues = New List(Of Integer)()
        End If
        Return items.ToMultiSelectListItems(name, value, selectedValues.[Select](Function(x) x.ToString()))
    End Function

    <Extension> _
    Public Function ToMultiSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValues As IEnumerable(Of Long)) As IEnumerable(Of SelectListItem)
        If selectedValues Is Nothing Then
            selectedValues = New List(Of Long)()
        End If
        Return items.ToMultiSelectListItems(name, value, selectedValues.[Select](Function(x) x.ToString()))
    End Function

    <Extension> _
    Public Function ToMultiSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValues As IEnumerable(Of Double)) As IEnumerable(Of SelectListItem)
        If selectedValues Is Nothing Then
            selectedValues = New List(Of Double)()
        End If
        Return items.ToMultiSelectListItems(name, value, selectedValues.[Select](Function(x) x.ToString()))
    End Function

    <Extension> _
    Public Function ToMultiSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValues As IEnumerable(Of Decimal)) As IEnumerable(Of SelectListItem)
        If selectedValues Is Nothing Then
            selectedValues = New List(Of Decimal)()
        End If
        Return items.ToMultiSelectListItems(name, value, selectedValues.[Select](Function(x) x.ToString()))
    End Function

    <System.Runtime.CompilerServices.Extension> _
    Public Function ToMultiSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), selectedValues As IEnumerable(Of String)) As IEnumerable(Of SelectListItem)
        If items Is Nothing Then
            Return New List(Of SelectListItem)()
        End If

        If selectedValues Is Nothing Then
            selectedValues = New List(Of String)()
        End If

        Return items.ToMultiSelectListItems(name, value, Function(x) selectedValues.Contains(value(x)))
    End Function

    <Extension> _
    Public Function ToMultiSelectListItems(Of T)(items As IEnumerable(Of T), name As Func(Of T, String), value As Func(Of T, String), isSelected As Func(Of T, Boolean)) As IEnumerable(Of SelectListItem)
        If items Is Nothing Then
            Return New List(Of SelectListItem)()
        End If

        Return items.[Select](Function(item) New SelectListItem() With { _
             .Text = name(item), _
             .Value = value(item), _
             .Selected = isSelected(item) _
        })
    End Function

End Module

File: Heat.ConvertedToC#/Extensions/EnumerableExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;

namespace Heat
{

	public static class EnumerableExtensions
	{

		public static IEnumerable<SelectListItem> ToSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, string selectedValue)
		{
			return items.ToSelectListItems(Name, value, selectedValue, false);
		}

		public static IEnumerable<SelectListItem> ToSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, string selectedValue, bool includeNotApplicable, string notApplicableValue = "", string notApplicableText = "(Not Applicable)")
		{
			return items.ToSelectListItems(Name, value, x => value(x) == selectedValue, includeNotApplicable, notApplicableValue, notApplicableText);
		}

		public static IEnumerable<SelectListItem> ToSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, Func<T, bool> isSelected)
		{
			return items.ToSelectListItems(Name, value, isSelected, false);
		}

		public static IEnumerable<SelectListItem> ToSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, Func<T, bool> isSelected, bool includeNotApplicable, string notApplicableValue = "", string notApplicableText = "(Not Applicable)")
		{
			if (includeNotApplicable) {
				var returnItems = new List<SelectListItem> { new SelectListItem {
					Text = notApplicableText,
					Value = notApplicableValue,
					Selected = false
				} };

				if (items != null) {
					returnItems.AddRange(items.Select(item => new SelectListItem {
						Text = Name(item),
						Value = value(item),
						Selected = isSelected(item)
					}));
				}
				return returnItems;
			}

			if (items == null) {
				return new List<SelectListItem>();
			}

			return items.Select(item => new SelectListItem {
				Text = Name(item),
				Value = value(item),
				Selected = isSelected(item)
			});
		}

		public static IEnumerable<SelectListItem> ToMultiSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, IEnumerable<int> selectedValues)
		{
			if (selectedValues == null) {
				selectedValues = new List<int>();
			}
			return items.ToMultiSelectListItems(Name, value, selectedValues.Select(x => x.ToString()));
		}

		public static IEnumerable<SelectListItem> ToMultiSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, IEnumerable<long> selectedValues)
		{
			if (selectedValues == null) {
				selectedValues = new List<long>();
			}
			return items.ToMultiSelectListItems(Name, value, selectedValues.Select(x => x.ToString()));
		}

		public static IEnumerable<SelectListItem> ToMultiSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, IEnumerable<double> selectedValues)
		{
			if (selectedValues == null) {
				selectedValues = new List<double>();
			}
			return items.ToMultiSelectListItems(Name, value, selectedValues.Select(x => x.ToString()));
		}

		public static IEnumerable<SelectListItem> ToMultiSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, IEnumerable<decimal> selectedValues)
		{
			if (selectedValues == null) {
				selectedValues = new List<decimal>();
			}
			return items.ToMultiSelectListItems(Name, value, selectedValues.Select(x => x.ToString()));
		}

		public static IEnumerable<SelectListItem> ToMultiSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, IEnumerable<string> selectedValues)
		{
			if (items == null) {
				return new List<SelectListItem>();
			}

			if (selectedValues == null) {
				selectedValues = new List<string>();
			}

			return items.ToMultiSelectListItems(Name, value, x => selectedValues.Contains(value(x)));
		}

		public static IEnumerable<SelectListItem> ToMultiSelectListItems<T>(this IEnumerable<T> items, Func<T, string> Name, Func<T, string> value, Func<T, bool> isSelected)
		{
			if (items == null) {
				return new List<SelectListItem>();
			}

			return items.Select(item => new SelectListItem {
				Text = Name(item),
				Value = value(item),
				Selected = isSelected(item)
			});
		}

	}
}

